quickwit/quickwit-search/src/top_k_collector.rs

```diff
@@ -0,0 +1,860 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::cmp::{Ordering, Reverse};
+use std::fmt::Debug;
+use std::marker::PhantomData;
+
+use quickwit_common::binary_heap::TopK;
+use quickwit_proto::search::{PartialHit, SortOrder};
+use tantivy::{DocId, Score};
+
+use crate::collector::{
+    HitSortingMapper, SegmentPartialHit, SegmentPartialHitSortingKey,
+    SortingFieldExtractorComponent, SortingFieldExtractorPair,
+};
+
+pub trait QuickwitSegmentTopKCollector {
+    fn collect_top_k_block(&mut self, docs: &[DocId]);
+    fn collect_top_k(&mut self, doc_id: DocId, score: Score);
+    fn get_top_k(&self) -> Vec<PartialHit>;
+}
+trait IntoOptionU64 {
+    fn is_unit_type() -> bool;
```

Comment:
nitpick. A default implementation might make sense here.
```suggestion
    fn is_unit_type() -> bool {
        false
    }
```